### ISSUE TITLE
feature request: show markdown messages in markdown, not plain text

### ISSUE THREAD

**Issue by lilmike** [ihabunek/toot#416]
Hiya,

I notice that when someone I follow posts a message using markdown, toot shows just the converted text content, which leaves out certain important things, such as links which simply show the text of the link, but not the link itself. Perhaps if the toot is written in markdown, the raw markdown should be displayed. This is using pleroma, I don't know if it's different on mastodon.

-Michael.

**Comment by danschwarz** [comment]
This gets tricky for two reasons:

1. I think, but need to verify, that Pleroma is using its own field, pleroma/content, alongside standard content in Status messages. Toot doesn't have special code to detect and show the pleroma/content field; it's just showing the regular content field.  Could this be changed? Yes.  _Should_ it be changed for benefit of Pleroma created content? Dunno.
2. Displaying just the text of a link, but not the link itself - I assume you mean showing the text as a clickable link like [this](http://www.example.com).  Well, that can be done, if and only if your terminal emulator supports the [OSC 8 standard](https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda).  We're adding support for that now in Master branch, with PR #415 just merged.  But there's no way for Toot to detect OCS 8 support itself, so we only use it in a limited fashion, by making existing http(s):// links clickable.

**Comment by lilmike** [comment]
Sorry, I wasn't super clear there: What I meant was, if you have
markdown code like this:

```
[this is a link](http://example.com)
```

Then toot will simply show 'this is a link' and the url will be nowhere
to be found.

-Michael.

**Comment by ihabunek** [comment]
I just tested and pleroma does not return the markdown representation of a toot even if it was posted in markdown.

The toot contained: `test links [bezdomni](https://bezdomni.net)`

Below is the json that Pleroma returns for the posted status. Notice that `content` is in HTML and `pleroma.content.text/plain` is plaintext. There is no markdown version.

I presume they convert the markdown (or bbcode which is also supported) into HTML and that's the canonical representation.

We could use something like [html2text](https://github.com/Alir3z4/html2text/) to convert back to markdown and show that when not using richtext.

```json
{
    "account": { ... },
    "application": null,
    "bookmarked": false,
    "card": null,
    "content": "<p>test links <a href=\"https://bezdomni.net\">bezdomni</a></p>",
    "created_at": "2023-11-19T11:18:51.000Z",
    "edited_at": null,
    "emojis": [],
    "favourited": false,
    "favourites_count": 0,
    "id": "AbxkRIkTDfzhXUQbPk",
    "in_reply_to_account_id": null,
    "in_reply_to_id": null,
    "language": null,
    "media_attachments": [],
    "mentions": [],
    "muted": false,
    "pinned": false,
    "pleroma": {
        "content": {
            "text/plain": "test links bezdomni"
        },
        "context": "https://localhost/contexts/357cbbca-1ac1-4ee7-a7ca-8bf81fbdbe0f",
        "conversation_id": 1631901516,
        "direct_conversation_id": null,
        "emoji_reactions": [],
        "expires_at": null,
        "in_reply_to_account_acct": null,
        "local": true,
        "parent_visible": false,
        "pinned_at": null,
        "quote": null,
        "quote_id": null,
        "quote_url": null,
        "quote_visible": false,
        "spoiler_text": {
            "text/plain": ""
        },
        "thread_muted": false
    },
    "poll": null,
    "reblog": null,
    "reblogged": false,
    "reblogs_count": 0,
    "replies_count": 0,
    "sensitive": false,
    "spoiler_text": "",
    "tags": [],
    "text": null,
    "uri": "https://localhost/objects/f4b0171c-fb9d-4ecd-9d8b-fc34c2633e0c",
    "url": "https://localhost/notice/AbxkRIkTDfzhXUQbPk",
    "visibility": "public"
}
```

**Comment by danschwarz** [comment]
> Sorry, I wasn't super clear there: What I meant was, if you have
> markdown code like this:
> 
> ```
> [this is a link](http://example.com)
> ```
> 
> Then toot will simply show 'this is a link' and the url will be nowhere
> to be found.
> 
> -Michael.
> 

But removal of the link would be a significant loss in content/functionality for the user.

**Comment by danschwarz** [comment]
> We could use something like [html2text](https://github.com/Alir3z4/html2text/) to convert back to markdown and show that when not using richtext.

Right, this is what I did to support rendering html messages to console or clipboard. I wish html2text was still maintained, but it's not. If we use it, we may need to take on the responsibility of keeping it working with subsequent Python releases, etc.

**Comment by lilmike** [comment]
Hiya,

See the previous reply for the reason toot isn't displaying the url, but
basically I was saying that if using pleroma, a toot rendered in
apparently html, originally written in markdown, will not show the
actual url. I was *not* suggesting that the url be removed, in fact
completely the opposite, I was simply stating that as it stands right
now, the url doesn't show up anywhere, and I only see the anchor text.
As replied, I suspect converting html back to markdown would be the best
way to go.

-Michael.

**Comment by danschwarz** [comment]
>  I was *not* suggesting that the url be removed, in fact completely the opposite, I was simply stating that as it stands right now, the url doesn't show up anywhere, and I only see the anchor text. As replied, I suspect converting html back to markdown would be the best way to go.

OK, thanks for clarifying, I must have misunderstood.
